crates/engine/tree/src/tree/payload_processor/prewarm.rs

```diff
@@ -78,37 +86,72 @@ where
     }
 
     /// Spawns all pending transactions as blocking tasks by first chunking them.
+    ///
+    /// For Optimism chains, special handling is applied to the first transaction if it's a
+    /// deposit transaction (type 0x7E/126) which sets critical metadata that affects all
+    /// subsequent transactions in the block.
     fn spawn_all(
         &self,
-        pending: mpsc::Receiver<impl ExecutableTxFor<Evm> + Send + 'static>,
+        pending: mpsc::Receiver<impl ExecutableTxFor<Evm> + Clone + Send + 'static>,
         actions_tx: Sender<PrewarmTaskEvent>,
     ) {
         let executor = self.executor.clone();
         let ctx = self.ctx.clone();
         let max_concurrency = self.max_concurrency;
+        let transaction_count_hint = self.transaction_count_hint;
 
         self.executor.spawn_blocking(move || {
-            let mut handles = Vec::with_capacity(max_concurrency);
             let (done_tx, done_rx) = mpsc::channel();
             let mut executing = 0;
-            while let Ok(executable) = pending.recv() {
-                let task_idx = executing % max_concurrency;
 
-                if handles.len() <= task_idx {
-                    let (tx, rx) = mpsc::channel();
-                    let sender = actions_tx.clone();
-                    let ctx = ctx.clone();
-                    let done_tx = done_tx.clone();
+            // Initially cap workers based on transaction count hint
+            let mut handles = Vec::with_capacity(max_concurrency);
+            let workers_needed = transaction_count_hint.min(max_concurrency);
 
-                    executor.spawn_blocking(move || {
-                        ctx.transact_batch(rx, sender, done_tx);
-                    });
+            // Spawn the required number of workers
+            for _ in 0..workers_needed {
```

Comment:
now that I see this, do we even configure the max concurrency anywhere, because spawning 64 tasks is probably not ideal

Reply:
i think its mostly hardcoded here as 64 here
https://github.com/paradigmxyz/reth/blob/yk/op-txn-prewarming/crates/engine/tree/src/tree/payload_processor/prewarm.rs#L89

Given that a normal txns prob average around 150 (mainnet) https://dune.com/queries/3149880 and base around 250 https://base.blockscout.com/blocks

16 would probably work?